DIPs/DIP1xxx-ajie.md

```diff
@@ -0,0 +1,144 @@
+as `auto ref`, which works for both referrable and non-referrable elements.
+
+This is to allow iteration over a range of non-copyable elements without explicit need
+to adapt the code for that.
+
+### Reference
+
+- [1] A pull request for DMD to disallow iteration by reference when the aggregate
+ does not support it:
+    * https://github.com/dlang/dmd/pull/8437
+
+- [2] Emsi-containers GitHub repository
+    * https://github.com/dlang-community/containers
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Alternatives](#alternatives)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+There are curently two ways to iterate over a range using foreach loop: By value and
+by reference.
+
+One is forced to iterate the range by reference if the element type is a `struct` with
+a disabled postblit, since iteration by value constructs the iteration variable by
+copy. As of DMD 2.082.0, you can iterate any range by reference, but there already
+is a pull request [1] to disallow such iteration for ranges which's front returns
+by value. That pull request is approved by Andrei Alexandrescu, implying that
+at least the underlying concept there is officially accepted.
+
+But in cases where the programmer does not modify the iteration variable, he/she does
+not care whether the iteration is done by reference or by value. Thus, there should
+be a way to iterate that works both with elements that are iterable only by
+value, and with non-copyable elements that can be iterated by reference.
+
+For function calls with a similar problem, there is already the `auto ref` keyword.
+For consistency, a similar keyword for `foreach` loop variables can be considered the
+best canditate to become such a way. On the other hand, if loop functionality excepts
+to be able to mutate the `foreach` aggreagate, existing `ref` could still be used,
+so the code will fail to compile if range elementscannot be mutated.
+
+Non-copyable `struct`s are an excellent aid when designing containers for RAII (Resource-
+Acquisition-Is-Initialization) principle. EMSI-Containers [2] are good examples of such
+non-copyable RAII containers. When one has many of those containers, it is natural to
+put them in a range. And with ranges comes the need to iterate.
+
+The iteration can be done in a general way with the current rules (see alternatives),
+or by always annotating/unannotating the iteration variable manually based on what is
+iterated over. But all of the former have their disadvantages, and the latter leads
+to needless changes in code when maintenance is done.
+
+## Description
+
+This DIP proposes, that when the compiler encounters a foreach statement:
+(example)
+
+```D
+foreach (auto ref loopVariable; aggregate)
+{
+    loopVariable.doSomething();
+}
+```
+
+...then if, and only if, elements of `aggregate` are lvalues, the compiler must treat the
```

Comment:
What do you mean by "elements of `aggregate` are lvalues"? I think that this should explained thoroughly and also the DIP must contain some real life examples highlighting the situations where `auto ref` is rewritten to both `ref` and nothing.

Reply:
Added a reference that explains lvalues (and rvalues).

As for the example, that's a good idea. I'll make a whole chapter for that -but not today.

Thanks for the review.